File: solutions/nim/perfect-numbers/1/perfect_numbers.nim
# Dedicated to Shree DR.MDD
import algorithm

type
  Classification* = enum
    Perfect, Abundant, Deficient

iterator getDivisors(val: int): int =
  for divr in 1 .. val div 2:
    if val mod divr == 0:
      yield divr

proc sumDivisors(val: int): int =
  for d in val.getDivisors:
    result.inc d

proc classify*(num: int): Classification =
  if num < 1:
    raise newException(ValueError, "Number not positive")
  let divSum = num.sumDivisors
  if divSum > num:
    Abundant
  elif divSum < num:
    Deficient
  else:
    Perfect

File: solutions/go/perfect-numbers/1/perfect_numbers.go
// Dedicated to Shree DR.MDD
package perfect

import (
	"errors"
)

type Classification int

var ErrOnlyPositive = errors.New("input is not a positive integer")

const (
	ClassificationDeficient Classification = iota
	ClassificationPerfect
	ClassificationAbundant
	ClassificationError
)

func Classify(number int64) (Classification, error) {
	if number <= 0 {
		return ClassificationError, ErrOnlyPositive
	}

	var sumFactors int64

	for i := int64(1); i <= number/2; i++ {
		if number%i == 0 {
			sumFactors += i
		}
	}

	switch {
	case number < sumFactors:
		return ClassificationAbundant, nil
	case number > sumFactors:
		return ClassificationDeficient, nil
	default:
		return ClassificationPerfect, nil
	}
}
